File: src/main/java/org/distril/beengine/server/ServerSettings.java
package org.distril.beengine.server;

import lombok.Getter;
import lombok.Setter;
import org.distril.beengine.config.Config;

import java.nio.file.Path;

@Getter
@Setter
public class ServerSettings {

	private final Config config;

	private String motd;
	private int maximumPlayers;
	private boolean xboxRequired;
	private boolean encryptionEnabled;
	private boolean debugEnabled;

	private String ip;
	private int port;
	private int compressionLevel;

	private String defaultWorldName;

	private int chunkExpiryTime;

	public ServerSettings(Path path) {
		this.config = Config.load(path.toFile());

		this.motd = this.config.get("server.motd", String.class);
		this.maximumPlayers = this.config.get("server.max-players", Integer.class);
		this.xboxRequired = this.config.get("server.xbox-require", Boolean.class);
		this.encryptionEnabled = this.config.get("server.enable-encryption", Boolean.class);
		this.debugEnabled = this.config.get("server.debug", Boolean.class);

		this.ip = this.config.get("network.ip", String.class);
		this.port = this.config.get("network.port", Integer.class);
		this.compressionLevel = this.config.get("network.compression-level", Integer.class);

		this.defaultWorldName = this.config.get("world.default", String.class);

		this.chunkExpiryTime = this.config.get("world.chunk.expiry-time", Integer.class);
	}

	public void save() {
		this.config.save();
	}
}

File: src/main/kotlin/org/distril/beengine/server/ServerSettings.kt
package org.distril.beengine.server

import org.spongepowered.configurate.ConfigurateException
import org.spongepowered.configurate.yaml.YamlConfigurationLoader
import java.io.IOException
import java.nio.file.Path

class ServerSettings(path: Path) {

	private val loader = YamlConfigurationLoader.builder().path(path).build()

	private val config by lazy {
		try {
			this.loader.load()
		} catch (exception: IOException) {
			throw RuntimeException("An error occurred while loading ServerSettings configuration:", exception)
		}
	}

	var motd = this.config.node("server", "motd").string!!
	var maximumPlayers = this.config.node("server", "max-players").int
	var isXboxRequired = this.config.node("server", "xbox-require").boolean
	var isEncryptionEnabled = this.config.node("server", "enable-encryption").boolean
	var isDebugEnabled = this.config.node("server", "debug").boolean

	var taskTimeout = this.config.node("scheduler", "timeout").long

	var ip = this.config.node("network", "ip").string!!
	var port = this.config.node("network", "port").int
	var compressionLevel = this.config.node("network", "compression-level").int

	var defaultWorldName = this.config.node("world", "default").string!!
	var chunkExpiryTime = this.config.node("world", "chunk", "expiry-time").int

	fun save() {
		try {
			this.loader.save(this.config)
		} catch (exception: ConfigurateException) {
			throw RuntimeException("Unable to save ServerSettings configuration: ${exception.message}")
		}
	}
}
